src/crowdin_client.cpp

```diff
@@ -57,56 +65,71 @@
 namespace
 {
 
-#define OAUTH_CLIENT_ID     "poedit-y1aBMmDbm3s164dtJ4Ur150e2"
-#define OAUTH_AUTHORIZE_URL "/oauth2/authorize?response_type=token&client_id=" OAUTH_CLIENT_ID
+#define OAUTH_SCOPE         "project+tm"
+#define OAUTH_CLIENT_ID     "k0uFz5HYQh0VzWgZmOpA"
+#define OAUTH_CLIENT_SECRET "p6Ko83cd4l4SD2TYeY8Fheffw8VuN3bML5jx7BeQ"
```

Comment:
The standing [strong recommendation](https://www.oauth.com/oauth2-servers/mobile-and-native-apps/) for native apps is to *not include client secret* — a native app like Poedit is a fundamentally insecure place for storing secrets (*especially* an Open Source one where it is published with zero obfuscation).

Instead, a different OAuth flow should be used, typically with server-side checking whether the callback URL matches the client. This is what Poedit does currently with v1 of Crowdin API.

After that change, `OAUTH_CLIENT_ID` will need to be re-generated and this one invalidated server side, because the (client_id, client_secret) pair has just been compromised by the publication of this PR.

Reply:
> The standing strong recommendation for native apps is to not include client secret — a native app like Poedit is a fundamentally insecure place for storing secrets (especially an Open Source one where it is published with zero obfuscation).

Fully agree, absolutely makes sense

> Instead, a different OAuth flow should be used, typically with server-side checking whether the callback URL matches the client. This is what Poedit does currently with v1 of Crowdin API.
> After that change, OAUTH_CLIENT_ID will need to be re-generated and this one invalidated server side, because the (client_id, client_secret) pair has just been compromised by the publication of this PR.

The problem here relaying on any extra online resource/service (even extremely simple taking minimal cost/resources on some shared hosting or cloud instance/resource) - creates extra potential point of failure and Poedit<=>Crowdin usage interruption, or extra surface for attack. And both above concerns seemed for me as more essential and higher priority over security. Plus security increased here relatively to what was before with addition of `refresh token` which becomes invalid every few hours and need to be refreshen. Though if you still find these security flaws critical - then how about ask Crowdin to implement what you expected on their side instead you. Since these secrets used only for their online service so IMO it would be more naturally to move responsibility for maintaining of that extra online resource/service (i.e. for keeping secret) to them while just freeing you from this :). Please let me know what do you think. Thanks

Reply:
> The problem here relaying on any extra online resource/service

Sorry, what extra services you mean? None are needed and Poedit already does it, i.e. uses [Implicit Flow](https://auth0.com/docs/flows/concepts/implicit). 

Current recommendation is to use [Authorization Code Flow with PKCE](https://auth0.com/docs/flows/concepts/auth-code-pkce) for native clients (which doesn't require any external services either), but that one is significantly more complicated one and the sole reason for its existence is to mitigate this attack:
* May make use of a custom URL scheme to capture redirects (e.g., poedit://) potentially allowing malicious applications to receive an Authorization Code from your Authorization Server.

This attack isn't a serious threat ([see also in-depth discussion](https://auth0.com/blog/oauth2-implicit-grant-and-spa/)) for two reasons:
1. Desktop OSes now prompt the user and ask if they should open a custom handler, and tell they user what app will open it — "Some Evil App" would be noticeable instead of the expected "Poedit".
2. With the exception of macOS, OS keychains can be accessed by other applications (please correct me if this is outdated). No need to intercept poedit:// handler when the token can be obtained from the keychain anyway. 

FWIW, I run api.poedit.net for Poedit Pro already, and the standard flow w/ client_secret could be securely passed through it, at extra complications and extra storage (for anonymous, account-less users, i.e. at volume). I wanted to avoid that precisely because of the Open Source version, which I think shouldn't rely on it, but also because of the extra burden. But it's a possibility.

What routing through api.poedit.net _would_ make much simpler, and more secure, is the Implicit Flow, because it would eliminate two annoyances: a) the desire for PKCE described above (URI handler intercept no longer possible) and b) the user-side hassle and associated failure-to-successfully-use with URI handlers. But that's a minor detail that can be decided later.

> Plus security increased here relatively to what was before with addition of refresh token which becomes invalid every few hours and need to be refreshen.

This is, at best, a false sense of security. In reality, there's no difference:
- with Implicit Flow, you have an access token stored on the client side that you can use to access the API at will
- with client_secret, you have a refresh_token stored on the client, plus know publicly disclosed client_secret, and you can use this to create an access token, then use it to access the API at will

You gained an extra step (and a lot of hassle in the legitimate client's code), but not any extra *security*.

> then how about ask Crowdin to implement what you expected on their side instead you

I would expect Implicit Flow to be supported out of the box by whatever library they use to implement OAuth 2.0. It's a standard thing.

Reply:
Thanks a lot for so extensive and exhaustive explanation. I am going get deeper into it and discuss with Crowdin within next couple of days. But just meantime answering of what I am fully confident.  

> Sorry, what extra services you mean?

I just mean that your server is involved in Poedit authorization process with that as it was implemented before and as you would expect to revert back to it now (so far I understand).
* So if your server would get down for any reason at least for some time (from which nobody can insure) - then nobody would be able to authorize with Poedit into Crowdin during your server downtime. What could never happen with current implementation in Poedit. 
* Or if you server would be compromised (from which alo nobody can insure) - then attacker would get much more chances to get access to much more Crowdin accounts then it could be with current implementation when attacker compromise authorization only for one user locally. 

> With the exception of macOS, OS keychains can be accessed by other applications (please correct me if this is outdated)

Reply:
> I just mean that your server is involved in Poedit authorization process with that **as it was implemented before**

That's not how it is currently implemented. Poedit uses Implicit Flow in the de-facto standard, run-of-the-mill way done by native & mobile apps, using a custom protocol in the callback URI:
1. When the user wants to login, it opens their default browser
2. The user logs into Crowdin and authorizes Poedit
3. Crowdin's page redirects to the OAuth callback URI, which is `poedit://auth/crowdin/<token_data>`
4. The browser reacts to this by showing a popup asking the user for permission to open the link in the Poedit app
5. Once the user agrees, Poedit opens, parses the URI and stores the token 

As you can see, Poedit's server doesn't enter the picture.

Reply:
Now I see that your server does **does not** receive `token` so does not get any access to customer's Crowdin account. Sorry for being misleading about this earlier. 

But while trying read carefully Implicit Flow by [link you provided above](https://auth0.com/docs/flows/concepts/implicit) and in [RFC as well](https://tools.ietf.org/html/rfc6749#section-1.3.2) - I still unable to understand:
* Why Poedit opens browser with URL https://poedit.net/crowdin?u=%2Foauth2%2Fauthorize%3Fresponse_type%3Dtoken%26client_id%3Dpoedit-y1aBMmDbm3s164dtJ4Ur150e2 on **your** domain ([poedit.net](https://poedit.net)) instead of   https://crowdin.com/oauth2/authorize?response_type=token&client_id=poedit-y1aBMmDbm3s164dtJ4Ur150e2 on **Crowdin's** domain ([crowdin.com](https://crowdin.com)) to which it's redirected from **your** website **right away anyway**. What is purpose of **mediation of your site** on exactly this step?
* When I press `Authorize` button on the page opened by above second URL (which is already on [crowdin.com](https://crowdin.com)) - does it redirects **right away** to **local Poedit URL handler** starting from [poedit://auth/crowdin/](poedit://auth/crowdin/?token=) and contained `token` already? Or to some URL on **your domain** yet from which redirected to [last above](poedit://auth/crowdin/?token=)? Seems already to [last above](poedit://auth/crowdin/?token=) so far I understand this step but just double checking.
* Implicit Flow [says here ](https://auth0.com/docs/flows/concepts/implicit#how-it-works) that tokens issued by OAuth server in this flow are _short-lived_ . Does it mean that in your implementation Poedit asks customer to sign in again every few hours (or whatever lifetime of `token` Crowdin's server defines) when it attempts to deal with Crowdin there? Since I just did not play with old OAuth in Crowdin a much enough to notice this.

Thanks a lot!

Reply:
What I think would be most productive would be if you downloaded Poedit 2.3 from https://poedit.net/download and tried the Crowdin workflow yourself.

> What is purpose of mediation of your site on exactly this step?

Crowdin's source tracking.

> does it redirects right away to local Poedit URL handler

Again, yes. Try it.

> Implicit Flow says here that tokens issued by OAuth server in this flow are short-lived

They are typically not, and they are not specifically in Crowdin's case.

Reply:
> Crowdin's source tracking.

If this is primary reason you would prefer to keep your OAuth implementation - we can discuss by email with Crowdin team alternative options you could get above info.

Reply:
> > Implicit Flow says here that tokens issued by OAuth server in this flow are short-lived

> They are typically not, and they are not specifically in Crowdin's case.

Yes, indeed in Crowdin's "Implicit flow" token lives 360 days (checked with them). So honestly speaking I am unable to see any technical/security advantages of your implementation with "Implicit Flow" over my current...except simplicity since agree that refresh token does not add any real security on Windows and Linux though still does on macOS (due of no access to secure storage from other apps under same user).

Reply:
> If this is primary reason you would prefer to keep your OAuth implementation

No, this is completely unrelated and the way you arrive (direct vs a redirect) at the first URL in the OAuth flow has no bearing on anything else in it.

> except simplicity

That's a pretty huge "except", but having client secret in native code, openly published, is also explicit A Very Bad Thing To Do by the spec, so I'm very reluctant to do it. 

> since agree that refresh token does not add any real security on Windows and Linux though still does on macOS 

It doesn't on macOS either, see above.

Reply:
Addressed in #633 2166b6102400a216d5853e5c0b78a621cdfc1e63 seems exactly as we discussed in this PR and also something by email. I.e. Implicit Flow is back with few refinements  described in above PR and commit message.